crates/red_knot/src/types.rs

```diff
@@ -0,0 +1,91 @@
+#![allow(dead_code)]
+use crate::Name;
+use ruff_index::{newtype_index, IndexVec};
+
+#[newtype_index]
+pub(crate) struct TypeId;
+
+impl TypeId {
+    pub(crate) fn typ(self, env: &TypeEnvironment) -> &Type {
+        env.type_for_id(self)
+    }
+}
+
+/// Arena holding all known types
+#[derive(Default)]
+pub(crate) struct TypeEnvironment {
+    types_by_id: IndexVec<TypeId, Type>,
+}
+
+impl TypeEnvironment {
+    pub(crate) fn add_class(&mut self, name: &str) -> TypeId {
+        let class = Type::Class(TypeClass {
+            name: Name::new(name),
+        });
+        self.types_by_id.push(class)
+    }
+
+    pub(crate) fn add_function(&mut self, name: &str) -> TypeId {
+        let function = Type::Function(TypeFunction {
+            name: Name::new(name),
+        });
+        self.types_by_id.push(function)
+    }
+
+    pub(crate) fn type_for_id(&self, type_id: TypeId) -> &Type {
+        &self.types_by_id[type_id]
+    }
+}
+
+pub(crate) enum Type {
+    Class(TypeClass),
+    Function(TypeFunction),
+}
+
+impl Type {
+    pub(crate) fn name(&self) -> &str {
+        match self {
+            Type::Class(inner) => inner.name(),
+            Type::Function(inner) => inner.name(),
+        }
+    }
+}
+
+pub(crate) struct TypeClass {
```

Comment:
Nit: I would rename this to `ClassType` because it's not a type class (a subset of a class) but instead it's the type representation of a class. 

Same for `TypeFunction`, rename to `FunctionType`. I think that also reads better with `LiteralType` rather than `TypeLiteral` (I think a type literal is something else, if it even exists).

Reply:
Yeah, I agree. I did this in parallel with how AST names its types, e.g. `Stmt::ClassDef` contains a `StmtClassDef`. But I don't like how it works in this case, I prefer `Type` as suffix also.